### PR TITLE
Fix: LogUnit FD/Memory leak

### PR DESCRIPTION
**Pull request by Maithem**
This patch evicts segment handles that correspond to addresses
that have been prefix trimmed. As a result of this eviction,
memory and open file descriptors are cleaned.

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 1]
```diff
@@ -1008,11 +1046,24 @@ public LogHeader getLogHeader() {
...
         private Set<Long> trimmedAddresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
         private Set<Long> pendingTrims = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        private volatile int refCount = 0;
 
 
+        public synchronized void retain() {
+            refCount++;
+        }
+
+        public synchronized void release() {
```
Comment (slfritchie) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: I've a question about the refcounting and `StreamLogFiles::close`, which isn't paying attention to the refcounts.  AFAICT, that method is only being used by unit test code, and honoring the refcounts in those tests probably doesn't matter?

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 2]
```diff
@@ -372,20 +379,37 @@ private void trimPrefix() {
...
 
-        for (File file : files) {
-            if (!file.delete()) {
-                log.error("Couldn't delete/trim file {}", file.getName());
+                File logFile = new File(sh.getFileName());
+                File trimmedFile = new File(getTrimmedFilePath(sh.getFileName()));
+                File pendingFile = new File(getPendingTrimsFilePath(sh.getFileName()));
+
+                if (!logFile.delete()) {
+                    log.error("Couldn't delete/trim file {}", logFile.getName());
```
Comment (no2chem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Style: trimPrefix: Couldn't ...

File: test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java [hunk 3]
```diff
@@ -417,6 +429,13 @@ public void testPrefixTrim() {
             }
         }
 
+        // Verify that the trimmed segment channels are close
```
Comment (no2chem) [inline test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java]: are closed

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 4]
```diff
@@ -937,22 +965,27 @@ public void append(long address, LogData entry) {
         } catch (IOException e) {
             log.error("Disk_write[{}]: Exception", address, e);
             throw new RuntimeException(e);
+        } finally {
```
Comment (medhavidhawan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: hopefully you have done the release in all places.

Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: I tested this with the l2 tests and lsof doesn't show any deleted entries. Let's see what happens on the other system

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java (hunk 2 above)
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Done.

File: test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java (hunk 3 above)
Comment (Maithem) [inline test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java]: Done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 5]
```diff
@@ -329,22 +331,27 @@ public void trim(long address) {
         SegmentHandle handle = getSegmentHandleForAddress(address);
         if (!handle.getKnownAddresses().containsKey(address)
                 || handle.getPendingTrims().contains(address)) {
+            handle.release();
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java (hunk 1 above)
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: You are right, I do that intentionally. The rational is, by the time the segment is closed, the ref count should be 0 and if it is not, then "forcefully" close the segment. Close will take precedence over waiting for the refcount to reach 0.

**Comment by Maithem** [comment]
Addressed all comments.

**Comment by Maithem** [comment]
Addresses https://github.com/CorfuDB/CorfuDB/pull/797

**Comment by Maithem** [comment]
@no2chem @medhavidhawan @slfritchie are you guys satisfied with the changes? can I merge?

**Pull request merged**